### PR TITLE
TESTS: added test for nested dlopen processing

### PR DESCRIPTION
**Pull request by hoopoepg**
- added test for processing of dl load sequence:
  exe->load lib1->load lib2
- it guarantee that all dynamic modules are processed by hook
  infrastructure

this is test for https://github.com/openucx/ucx/pull/2898

### CODE REVIEW

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/7818/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/7819/ for details (Mellanox internal link).

File: test/gtest/ucm/malloc_hook.cc [hunk 1]
```diff
@@ -799,3 +800,78 @@ UCS_TEST_F(malloc_hook, bistro_patch) {
...
 }
+
+/* test for mmap events are fired from non-direct load modules
+ * we are trying to load lib1, from lib1 load lib2, and
+ * fire mmap event from lib2 */
+UCS_TEST_F(malloc_hook, dlopen) {
+    typedef void (fire_mmap_f)(void);
+    typedef void* (load_lib_f)(const char *path);
+
+    const char *libdlopen_test  = "/.libs/libdlopen_test.so";
```
Comment (yosefe) [inline test/gtest/ucm/malloc_hook.cc]: looks a bit ugly to use paths like this in gtest..
maybe do it in separate test in test_jenkins.sh (with mpi malloc hook tests)?

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/7826/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/7827/ for details (Mellanox internal link).

File: test/gtest/Makefile.am [hunk 2]
```diff
@@ -55,7 +57,8 @@ gtest_CPPFLAGS = \
 
 gtest_LDFLAGS  = $(GTEST_LDFLAGS) -no-install
 gtest_CFLAGS   = $(BASE_CFLAGS)
-gtest_CXXFLAGS = $(BASE_CXXFLAGS) $(GTEST_CXXFLAGS) -g -fno-tree-vectorize
+gtest_CXXFLAGS = $(BASE_CXXFLAGS) $(GTEST_CXXFLAGS) -g -fno-tree-vectorize \
+				 -DGTEST_UCM_HOOK_LIB_DIR="\"${PWD}/ucm/test_dlopen/.libs\""
```
Comment (yosefe) [inline test/gtest/Makefile.am]: pls use automake vars (e.g abs_builddir) instead of PWD

File: test/gtest/ucm/malloc_hook.cc [hunk 3]
```diff
@@ -799,3 +800,77 @@ UCS_TEST_F(malloc_hook, bistro_patch) {
     EXPECT_NE(patched, origin);
 #endif
 }
+
+/* test for mmap events are fired from non-direct load modules
+ * we are trying to load lib1, from lib1 load lib2, and
+ * fire mmap event from lib2 */
+UCS_TEST_F(malloc_hook, dlopen) {
+#ifndef GTEST_UCM_HOOK_LIB_DIR
+    UCS_TEST_SKIP_R("missing build configuration");
```
Comment (yosefe) [inline test/gtest/ucm/malloc_hook.cc]: better fail to compile, than to skip the test

File: test/gtest/ucm/malloc_hook.cc [hunk 4]
```diff
@@ -799,3 +800,77 @@ UCS_TEST_F(malloc_hook, bistro_patch) {
...
+    ucs_status_t status;
+
+    status = ucm_set_event_handler(UCM_EVENT_VM_MAPPED,
+                                   0, mem_event_callback,
+                                   reinterpret_cast<void*>(this));
+    ASSERT_UCS_OK(status);
+
+    lib_name  = (char*)malloc(strlen(GTEST_UCM_HOOK_LIB_DIR) + strlen(libdlopen_test) + 1);
+    lib2_name = (char*)malloc(strlen(GTEST_UCM_HOOK_LIB_DIR) + strlen(libdlopen_test2) + 1);
+    strcat(strcpy(lib_name, GTEST_UCM_HOOK_LIB_DIR), libdlopen_test);
```
Comment (yosefe) [inline test/gtest/ucm/malloc_hook.cc]: we have std::string to to string manipulations (operator+), no need to work so hard with strcat and malloc

File: test/gtest/ucm/malloc_hook.cc [hunk 5]
```diff
@@ -799,3 +800,77 @@ UCS_TEST_F(malloc_hook, bistro_patch) {
...
+no_fire:
+    dlclose(lib2);
+no_load:
+    dlclose(lib);
+no_lib:
+    free(lib_name);
+    free(lib2_name);
+    ucm_unset_event_handler(UCM_EVENT_VM_MAPPED, mem_event_callback,
+                            reinterpret_cast<void*>(this));
+    return;
```
Comment (yosefe) [inline test/gtest/ucm/malloc_hook.cc]: why need return?

File: test/gtest/ucm/test_dlopen/Makefile.am [hunk 6]
```diff
@@ -0,0 +1,16 @@
+#
+# Copyright (C) Mellanox Technologies Ltd. 2001-2014.  ALL RIGHTS RESERVED.
+#
+# See file LICENSE for terms.
+#
+
+
+noinst_lib_LTLIBRARIES = \
```
Comment (yosefe) [inline test/gtest/ucm/test_dlopen/Makefile.am]: shouldn't it be noinst_LTLIBRARIES ?

File: test/gtest/ucm/test_dlopen/Makefile.am [hunk 7]
```diff
@@ -0,0 +1,16 @@
+#
+# Copyright (C) Mellanox Technologies Ltd. 2001-2014.  ALL RIGHTS RESERVED.
```
Comment (yosefe) [inline test/gtest/ucm/test_dlopen/Makefile.am]: 2018

File: test/gtest/ucm/test_dlopen/Makefile.am [hunk 8]
```diff
@@ -0,0 +1,16 @@
...
+# See file LICENSE for terms.
+#
+
+
+noinst_lib_LTLIBRARIES = \
+	libdlopen_test.la \
+	libdlopen_test2.la
+
+libdlopen_test_la_SOURCES   = libtest_dlopen.c
+libdlopen_test2_la_SOURCES  = libtest_dlopen2.c
```
Comment (yosefe) [inline test/gtest/ucm/test_dlopen/Makefile.am]: maybe the could be more meaningful name than dlopen_test2?
e.g dlopen_test_do_load.so, dlopen_test_do_mmap.so

File: test/gtest/ucm/test_dlopen/libtest_dlopen2.c [hunk 9]
```diff
@@ -0,0 +1,10 @@
+#include <sys/mman.h>
+#include <errno.h>
```
Comment (yosefe) [inline test/gtest/ucm/test_dlopen/libtest_dlopen2.c]: why need errno.h?

File: test/gtest/ucm/test_dlopen/libtest_dlopen.c [hunk 10]
```diff
@@ -0,0 +1,6 @@
+#include <dlfcn.h>
```
Comment (yosefe) [inline test/gtest/ucm/test_dlopen/libtest_dlopen.c]: pls add copyright

File: test/gtest/ucm/test_dlopen/libtest_dlopen2.c [hunk 11]
```diff
@@ -0,0 +1,10 @@
+#include <sys/mman.h>
```
Comment (yosefe) [inline test/gtest/ucm/test_dlopen/libtest_dlopen2.c]: pls add copyright

File: test/gtest/ucm/test_dlopen/Makefile.am (hunk 7 above)
Comment (hoopoepg) [reply test/gtest/ucm/test_dlopen/Makefile.am]: fixed

File: test/gtest/ucm/test_dlopen/Makefile.am (hunk 6 above)
Comment (hoopoepg) [reply test/gtest/ucm/test_dlopen/Makefile.am]: noinst_LTLIBRARIES doesn't work

File: test/gtest/Makefile.am [hunk 12]
```diff
@@ -55,7 +57,8 @@ gtest_CPPFLAGS = \
 
 gtest_LDFLAGS  = $(GTEST_LDFLAGS) -no-install
 gtest_CFLAGS   = $(BASE_CFLAGS)
-gtest_CXXFLAGS = $(BASE_CXXFLAGS) $(GTEST_CXXFLAGS) -g -fno-tree-vectorize
+gtest_CXXFLAGS = $(BASE_CXXFLAGS) $(GTEST_CXXFLAGS) -g -fno-tree-vectorize \
```
[the remark this answers is not in the document]
Comment (yosefe) [reply test/gtest/Makefile.am]: if this is really the case, need to fix BASE_CXXFLAGS

[the remark this answers is not in the document]
Comment (hoopoepg) [reply test/gtest/Makefile.am]: hmmm, BASE_CXXFLAGS inherits BASE_CFLAGS, which is defined as
```
BASE_CFLAGS="-g -Wall -Werror"
```
it means that -g is always used to build UCX... is it correct?

[the remark this answers is not in the document]
Comment (yosefe) [reply test/gtest/Makefile.am]: yes

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/7860/ for details (Mellanox internal link).

File: test/gtest/ucm/malloc_hook.cc [hunk 13]
```diff
@@ -829,12 +829,13 @@ UCS_TEST_F(malloc_hook, dlopen) {
...
 
-    lib_name  = (char*)malloc(strlen(GTEST_UCM_HOOK_LIB_DIR) + strlen(libdlopen_test) + 1);
-    lib2_name = (char*)malloc(strlen(GTEST_UCM_HOOK_LIB_DIR) + strlen(libdlopen_test2) + 1);
-    strcat(strcpy(lib_name, GTEST_UCM_HOOK_LIB_DIR), libdlopen_test);
-    strcat(strcpy(lib2_name, GTEST_UCM_HOOK_LIB_DIR), libdlopen_test2);
+    lib_load = std::string(GTEST_UCM_HOOK_LIB_DIR) + libdlopen_load;
+    lib_mmap = std::string(GTEST_UCM_HOOK_LIB_DIR) + libdlopen_mmap;
 
-    lib = dlopen(lib_name, RTLD_NOW);
+    UCS_TEST_MESSAGE << lib_load;
```
Comment (yosefe) [inline test/gtest/ucm/malloc_hook.cc]: need to print also what it is (e.g "Loading " << lib_load), or remove the message

File: test/gtest/Makefile.am (hunk 2 above)
Comment (yosefe) [reply test/gtest/Makefile.am]: looks like it's not fixed..

File: test/gtest/ucm/malloc_hook.cc (hunk 13 above)
Comment (hoopoepg) [reply test/gtest/ucm/malloc_hook.cc]: added "Loading"

File: test/gtest/Makefile.am (hunk 2 above)
Comment (hoopoepg) [reply test/gtest/Makefile.am]: removed -g and missed about PWD :)
fixed

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/7882/ for details (Mellanox internal link).

**Comment by yosefe** [comment]
looks good except need to address https://github.com/openucx/ucx/pull/2904#discussion_r222608166

**Comment by hoopoepg** [comment]
yep, missed few commits when pushed changes...
fixed again

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/7886/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/7889/ for details (Mellanox internal link).

**Pull request merged**